### PR TITLE
Async-streams: Fix crash with async-iterator method with only throw

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/31552

For context, a method body without a `yield` is not considered an iterator.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.cs [hunk 1]
```diff
@@ -47,6 +47,16 @@ internal partial class AsyncRewriter : StateMachineRewriter
                 return body;
             }
 
+            CSharpCompilation compilation = method.DeclaringCompilation;
+            bool isAsyncEnumerableOrEnumerator = method.IsIAsyncEnumerableReturningAsync(compilation) ||
+                method.IsIAsyncEnumeratorReturningAsync(compilation);
+            if (isAsyncEnumerableOrEnumerator && !method.IsIterator)
+            {
+                diagnostics.Add(new CSDiagnosticInfo(ErrorCode.ERR_PossibleAsyncIteratorWithoutYield, method.ReturnType), method.Locations[0]);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.cs]: >diagnostics.Add(new CSDiagnosticInfo(ErrorCode.ERR_PossibleAsyncIteratorWithoutYield, method.ReturnType), method.Locations[0]) [](start = 16, length = 126)

Consider using extension method:
```C#
diagnostics.Add(ErrorCode.ERR_PossibleAsyncIteratorWithoutYield, method.Locations[0], method.ReturnType)
```

**Comment by 333fred** [comment]
Done review pass (commit 2)

**Comment by jcouv** [comment]
LDM decided to make this scenario an error, as implemented in this PR. So to be an async-iterator, you must be `async`, you must return an acceptable type, and you must be an iterator (ie. have a `yield` of some kind).
But they suggested that the error message should de-emphasize "use a different type" as the recommendation. Instead, the message should focus on (1) removing `async`, or (2) adding `yield`.

I'll update the PR.

**Comment by jcouv** [comment]
@333fred Please take another look when you have some time. Thanks

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs [hunk 2]
```diff
@@ -1205,6 +1206,161 @@ static async System.Collections.Generic.IAsyncEnumerator<int> M(int value)
...
+                //     async System.Collections.Generic.IAsyncEnumerator<int> M()
+                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "M").WithLocation(4, 60)
+                );
+            comp.VerifyEmitDiagnostics(
+                // (4,60): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
+                //     async System.Collections.Generic.IAsyncEnumerator<int> M()
+                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "M").WithLocation(4, 60),
+                // (4,60): error CS8420: The body of an async-iterator method must contain a 'yield' statement. Consider removing `async` from the method declaration.
+                //     async System.Collections.Generic.IAsyncEnumerator<int> M()
+                Diagnostic(ErrorCode.ERR_PossibleAsyncIteratorWithoutYieldOrAwait, "M").WithArguments("System.Collections.Generic.IAsyncEnumerator<int>").WithLocation(4, 60)
```
Comment (alrz) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs]: How come this is an error rather than a warning?

Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs]: @alrz Imagine that the `IAsyncEnumerable` type was tagged to make it a task-like attribute. We don't know whether to emit a state machine for async method, or for async-iterator method.
If we can't compile, then the diagnostic must be an error.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 3]
```diff
@@ -2641,6 +2641,12 @@ A catch() block after a catch (System.Exception e) block can catch non-CLS excep
   <data name="ERR_AwaitForEachMissingMemberWrongAsync" xml:space="preserve">
     <value>Asynchronous foreach statement cannot operate on variables of type '{0}' because '{0}' does not contain a public instance definition for '{1}'. Did you mean 'foreach' rather than 'await foreach'?</value>
   </data>
+  <data name="ERR_PossibleAsyncIteratorWithoutYield" xml:space="preserve">
+    <value>The body of an async-iterator method must contain a 'yield' statement.</value>
+  </data>
+  <data name="ERR_PossibleAsyncIteratorWithoutYieldOrAwait" xml:space="preserve">
+    <value>The body of an async-iterator method must contain a 'yield' statement. Consider removing `async` from the method declaration.</value>
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: >Consider removing `async` from the method declaration. [](start = 82, length = 54)

Nit: consider changing to `Consider adding a 'yield' statement or removing 'async' from the method declaration.

**Pull request merged**